core/node/provider.go

```diff
@@ -40,88 +41,100 @@ const reprovideStrategyKey = "/reprovideStrategy"
-
-type Provider interface {
-	// StartProviding provides the given keys to the DHT swarm unless they were
-	// already provided in the past. The keys will be periodically reprovided until
-	// StopProviding is called for the same keys or user defined garbage collection
-	// deletes the keys.
-	StartProviding(...mh.Multihash)
+func (r *NoopProvider) StartProviding(bool, ...mh.Multihash) {}
+func (r *NoopProvider) StopProviding(...mh.Multihash)        {}
+func (r *NoopProvider) ProvideOnce(...mh.Multihash)          {}
+func (r *NoopProvider) ClearProvideQueue() int               { return 0 }
+
+// DHTProvider is an interface for providing keys to a DHT swarm. It holds a
+// state of keys to be advertised, and is responsible for periodically
+// publishing provider records for these keys to the DHT swarm before the
+// records expire.
+type DHTProvider interface {
+	// StartProviding ensures keys are periodically advertised to the DHT swarm.
+	//
+	// If the `keys` aren't currently being reprovided, they are added to the
+	// queue to be provided to the DHT swarm as soon as possible, and scheduled
+	// to be reprovided periodically. If `force` is set to true, all keys are
+	// provided to the DHT swarm, regardless of whether they were already being
+	// reprovided in the past. `keys` keep being reprovided until `StopProviding`
+	// is called.
+	//
+	// This operation is asynchronous, it returns as soon as the `keys` are added
+	// to the provide queue, and provides happens asynchronously.
+	StartProviding(force bool, keys ...mh.Multihash)
 
 	// StopProviding stops reproviding the given keys to the DHT swarm. The node
 	// stops being referred as a provider when the provider records in the DHT
 	// swarm expire.
-	StopProviding(...mh.Multihash)
-
-	// InstantProvide only sends provider records for the given keys out to the DHT
-	// swarm. It does NOT take the responsibility to reprovide these keys.
-	InstantProvide(context.Context, ...mh.Multihash) error
-
-	// ForceProvide is similar to StartProviding, but it sends provider records out
-	// to the DHT even if the keys were already provided in the past.
-	ForceProvide(context.Context, ...mh.Multihash) error
+	//
+	// Remove the `keys` from the schedule and return immediately. Valid records
+	// can remain in the DHT swarm up to the provider record TTL after calling
+	// `StopProviding`.
+	StopProviding(keys ...mh.Multihash)
 
-	// TODO: Do we need this?
+	// ProvideOnce sends provider records for the specified keys to the DHT swarm
+	// only once. It does not automatically reprovide those keys afterward.
 	//
-	// Remove and keys that are being reprovided.
-	Clear() int
+	// Add the supplied multihashes to the provide queue, and return immediately.
+	// The provide operation happens asynchronously.
+	ProvideOnce(keys ...mh.Multihash)
+
+	// ClearProvideQueue clears the all the keys from the provide queue and returns
+	// the number of keys that were cleared.
+	ClearProvideQueue() int
```

Comment:
Comment should make it clear whether or not this clears out the keystore also. 

I assume this does not clear the keystore, in which case we also nees a `Purge` function that clears everything.